dandiapi/api/services/metadata/__init__.py

```diff
@@ -62,14 +59,20 @@ def validate_asset_metadata(*, asset: Asset) -> None:
             asset.status = Asset.Status.INVALID
             asset.validation_errors = [{'field': '', 'message': str(e)}]
 
-        # Save asset
-        asset.save()
-
-        # Update modified timestamps on all draft versions this asset belongs to
-        asset.versions.filter(version='draft').update(modified=timezone.now())
+        updated_asset = Asset.objects.filter(
+            id=asset.id, status=Asset.Status.PENDING, metadata=asset.metadata, published=False
+        ).update(
+            status=asset.status,
+            validation_errors=asset.validation_errors,
+            # include metadata in update since we're bypassing .save()
+            metadata=asset._populate_metadata(),
+        )
+        if updated_asset:
+            # Update modified timestamps on all draft versions this asset belongs to
+            asset.versions.filter(version='draft').update(modified=timezone.now())
```

Comment:
What's the reason for bypassing `asset.save()` here?

Reply:
In general I think we need to move away from `.save` unless we're using `select_for_update`. Otherwise modifications could come in between the time this function starts and completes and they would be immediately overwritten, similar to #1557. Here I'm using filter criteria of `id,status,metadata,published` to give us a version of optimistic locking that's pretty robust to that failure mode. The update would no-op and the next tick of the cron would attempt it again.

Reply:
> In general I think we need to move away from `.save` unless we're using `select_for_update`. 

1000% agree on this ^^

Reply:
> In general I think we need to move away from .save unless we're using select_for_update. Otherwise modifications could come in between the time this function starts and completes and they would be immediately overwritten, similar to https://github.com/dandi/dandi-archive/pull/1557.

Great point, yeah I agree with that. 

> Here I'm using filter criteria of id,status,metadata,published to give us a version of optimistic locking that's pretty robust to that failure mode. The update would no-op and the next tick of the cron would attempt it again.

There's something about this that doesn't sit quite right with me. This function used to accept an asset in any state, run validation on it, and save the result. Now, it will accept an asset in any state, run validation on it, but then only save the result if the status is `PENDING`. I feel like the silent failure mode is asking for trouble. Maybe we should raise an error at the beginning of the function, similar to what we do for `asset.published`? Or maybe just an if statement that logs the fact that we're skipping this asset because it's pending. What do you think?

Reply:
> This function used to accept an asset in any state, run validation on it, and save the result. Now, it will accept an asset in any state, run validation on it, but then only save the result if the status is PENDING.

This is just a mistake on my part and overfitting this to the cron use case. The `revalidate` script as it is would silently fail right now. I tweaked this to accept an asset in any state. I also added a log message about the asset being modified while it was validating. 

I think part of problem here is that we need to overly specify our `filter` because we're using it as a proxy to mean "only update the asset if it hasn't been modified". If we were using a revision column similar to how https://github.com/saxix/django-concurrency works, then the update query would only filter on the `id,revision` pair plus any additional requirements (in this case `published`).

Even with the changes made, it's still possible for a caller to get a log message and nothing else to indicate that the call failed. Maybe this is as simple as returning a boolean. In the case of the celery beat, we ignore it because we don't need it, but something like the revalidate script might want to retry, abort, or something else. What do you think?

Reply:
I like that pattern, and it looks like you've already done that here, so this looks good to me.